Example of using perf_analyzer with DALI backend (#28)
triton-inference-server/dali_backend: docs/examples/perf_analyzer/README.md

# perf_analyzer example with DALI image decoder

## Run the example

### tritonserver

To run the tritonserver, first you need to serialize DALI pipeline and put it inside model repo.

`setup_perf_analyzer_example.sh` is a convenience script, that automatizes setting up.

> Provided you have DALI wheel installed in your system, you can just call `sh ./setup_perf_analyzer_example.sh`

Author: I retained the `sh setup_perf_analyzer_example.sh`. I don't want to set `+x` mod for the `setup...sh` file. So that's the way you should actually run it.

When you have your model repo set up, you can run `tritonserver`. Be sure to fill `<path to model repo>` properly:

>     docker run -it --rm --shm-size=1g --ulimit memlock=-1 --gpus all --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 --privileged -v <path to model repo>:/models nvcr.io/nvidia/tritonserver:XX.YY-py3 tritonserver --model-repository=/models

Reviewer: Can you provide any working example that the user can just copy-past and run (example values for XX.YY and <path to model repo>)?

Author: I don't really know, what `-v <path to model repo>:/models` user will have. As far as `XX.YY` goes, I used it so that this doc won't need to be updated every time new `tritonserver` version is released.

Reviewer: I was asking for a simple example that anyone can just copy-past and run just to check how it works before moving to the real profiling.
Also I don't really know what `<path to model repo>:/models` should be. So even a simple example would provide a lot of info.

Author: To run this example, user has to set up the model repository quasi-manually (using the `setup` script, it's mentioned earlier in the README). Therefore I wouldn't like to impose any model_repository path here. I've put a short explanation there, following @jantonguirao suggestion. Is that ok?

Reviewer: Sure. I'm just saying that I don't know TRION much and I cannot run this easily on my own to verify.

>     docker run -it --rm --shm-size=1g --ulimit memlock=-1 --gpus all --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 --privileged -v <path to model repo>:/models nvcr.io/nvidia/tritonserver:XX.YY-py3 tritonserver --model-repository=/models

Reviewer: ```suggestion
    docker run -it --rm --shm-size=1g --ulimit memlock=-1 --gpus all --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 -v <path to model repo>:/models nvcr.io/nvidia/tritonserver:XX.YY-py3 tritonserver --model-repository=/models
```

>     docker run -it --rm --shm-size=1g --ulimit memlock=-1 --gpus all --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 --privileged -v <path to model repo>:/models nvcr.io/nvidia/tritonserver:XX.YY-py3 tritonserver --model-repository=/models

Author: This is actually pretty standard way of running `tritonserver`. I believe that if user is reading this, he for sure already knows it. Here's the reference material: [Triton quickstart](https://github.com/triton-inference-server/server/blob/master/docs/quickstart.md#run-on-system-with-gpus)

>     docker run -it --rm --shm-size=1g --ulimit memlock=-1 --gpus all --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 --privileged -v <path to model repo>:/models nvcr.io/nvidia/tritonserver:XX.YY-py3 tritonserver --model-repository=/models

Reviewer: Maybe you can mention that this follow the Triton quickstart guide?

>     docker run -it --rm --shm-size=1g --ulimit memlock=-1 --gpus all --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 --privileged -v <path to model repo>:/models nvcr.io/nvidia/tritonserver:XX.YY-py3 tritonserver --model-repository=/models

Author: Added a line on top of the README file

### perf_analyzer

> 1. Pick an image you like the most

Reviewer: ```suggestion
1. Pick an image for testing
```

1. Run Triton's client docker container:
 
    ```
    docker run -it --net=host nvcr.io/nvidia/tritonserver:XX.YY-py3-clientsdk
    ```

1. Create a directory for the test image and put it there. 
**IMPORTANT**: file must have the same name as the input (`DALI_INPUT_0` in our case)
    
    ```
    mkdir test_image

>     touch test_image/DALI_INPUT_0  # it's not really "touch" - it's an image

Reviewer: Why you need to `touch` it?

Author: As I wrote in the comment, it's not really `touch`, it's an image. I thought that since I create this directory a line above, it's be obvious. Is it unclear?

Reviewer: So how about:
```
cp PATH_TO_YOUR_IMAGE test_image/DALI_INPUT_0
```

Author: Done

# perf_analyzer example with DALI image decoder

This is an example of using `perf_analyer` tool from the Triton's Client toolkit in order to
measure the performance of your model ensemble, when DALI is the first model in the ensemble.
Before diving into this example, be sure to get accustomed to [Triton's basics](https://github.com/triton-inference-server/server#documentation).

## Run the example

### tritonserver

To run the tritonserver, first you need to serialize a DALI pipeline and put it inside the model repository.

`setup_perf_analyzer_example.sh` is a convenience script that automatizes setting up.
Provided you have DALI installed in your system, you can just call `sh setup_perf_analyzer_example.sh`

> When you have your model repository set up, you can run `tritonserver`. Be sure to replace `<path to model repo>` with the actual path in your system:

Reviewer: ```suggestion
When you have your model repository set up, you can run `tritonserver`. Be sure to replace `<path to model repo>` with the actual path in your system (`21.02-py3 tritonserver` version is just an example, any newer should work as well):
```

    docker run -it --rm --shm-size=1g --ulimit memlock=-1 --gpus all --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 -v <path to model repo>:/models nvcr.io/nvidia/tritonserver:21.02-py3 tritonserver --model-repository=/models

### perf_analyzer

1. Pick an image for testing

> 1. Run Triton's client docker container:
>  
>     ```
>     docker run -it --net=host nvcr.io/nvidia/tritonserver:20.11-py3-clientsdk
>     ```
> 

Reviewer: ```suggestion
1. Run Triton's client docker container (`20.11-py3-clientsdk` version is just an example, any newer should work as well):
 
    ```
    docker run -it --net=host nvcr.io/nvidia/tritonserver:20.11-py3-clientsdk
    ```

```
